circuits/src/recipient.rs

```diff
@@ -0,0 +1,114 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_jubjub::JubJubScalar;
+use dusk_plonk::prelude::*;
+use jubjub_schnorr::{gadgets, Signature};
+
+use crate::elgamal;
+use phoenix_core::PublicKey;
+
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+use phoenix_core::SecretKey;
+
+const TX_OUTPUT_NOTES: usize = 2;
+
+/// Parameters needed to prove a recipient in-circuit
+#[derive(Debug, Clone, Copy)]
+pub struct RecipientParameters {
+    /// Public key of the transaction sender
+    pub sender_pk: PublicKey,
+    /// Note public keys of each note recipient
+    pub recipient_npk_vec: [JubJubAffine; TX_OUTPUT_NOTES],
+    /// Signatures of 'payload_hash' verifiable using 'pk_A' and 'pk_B'
+    pub sig_vec: [Signature; TX_OUTPUT_NOTES],
+    /// Asymmetric encryption of 'pk_A' using both recipients 'npk'
+    pub enc_A_vec: [(JubJubExtended, JubJubExtended); TX_OUTPUT_NOTES],
+    /// Asymmetric encryption of 'pk_B' using both recipients 'npk'
+    pub enc_B_vec: [(JubJubExtended, JubJubExtended); TX_OUTPUT_NOTES],
+    /// Randomness needed to encrypt/decrypt 'pk_A'
+    pub r_A_vec: [JubJubScalar; TX_OUTPUT_NOTES],
+    /// Randomness needed to encrypt/decrypt 'pk_B'
+    pub r_B_vec: [JubJubScalar; TX_OUTPUT_NOTES],
+}
+
+impl Default for RecipientParameters {
+    fn default() -> Self {
+        let mut rng = StdRng::seed_from_u64(0xbeef);
+
+        let sk = SecretKey::random(&mut rng);
+        let sender_pk = PublicKey::from(&sk);
+
+        Self {
+            sender_pk,
+            recipient_npk_vec: [
+                JubJubAffine::default(),
+                JubJubAffine::default(),
+            ],
+            sig_vec: [Signature::default(), Signature::default()],
+            enc_A_vec: [(JubJubExtended::default(), JubJubExtended::default());
+                TX_OUTPUT_NOTES],
+            enc_B_vec: [(JubJubExtended::default(), JubJubExtended::default());
+                TX_OUTPUT_NOTES],
+            r_A_vec: [JubJubScalar::default(); TX_OUTPUT_NOTES],
+            r_B_vec: [JubJubScalar::default(); TX_OUTPUT_NOTES],
+        }
+    }
+}
+
+/// Gadget to prove a valid origin for a given transaction.
+pub(crate) fn gadget(
+    composer: &mut Composer,
+    rp: &RecipientParameters,
+    payload_hash: &Witness,
```

Comment:
Witnesses are copy, so just take them by value. They're just a `usize` anyway.
```suggestion
    payload_hash: Witness,
```